### PR TITLE
Remove support for internal versioning for concurrency control

### PR DESCRIPTION
**Pull request by bleskes**
Elasticsearch has long [supported](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-index_.html#index-versioning) compare and set (a.k.a optimistic concurrency control) operations using internal document versioning. Sadly that approach is flawed and can sometime do the wrong thing. Here's the relevant excerpt from the resiliency status page:

> When a primary has been partitioned away from the cluster there is a short period of time until it detects this. During that time it will continue indexing writes locally, thereby updating document versions. When it tries to replicate the operation, however, it will discover that it is partitioned away. It won’t acknowledge the write and will wait until the partition is resolved to negotiate with the master on how to proceed. The master will decide to either fail any replicas which failed to index the operations on the primary or tell the primary that it has to step down because a new primary has been chosen in the meantime. Since the old primary has already written documents, clients may already have read from the old primary before it shuts itself down. The version numbers of these reads may not be unique if the new primary has already accepted writes for the same document 

We recently [introduced](https://www.elastic.co/guide/en/elasticsearch/reference/6.x/optimistic-concurrency-control.html) a new sequence number based approach that doesn't suffer from this dirty reads problem. 

This PR removes support for internal versioning as a concurrency control mechanism in favor of the sequence number approach.

*Notes*: 
1) I'm opening this PR to allow downstream projects to test their code to see it doesn't use versioning.
2) I will open a separate PR with deprecation warnings for 6.x
3) I still need to add breaking changes documentation
4) This PR contain some minor fixes I run into that I will extract to a separate PR before merging.

Relates to #1078

### CODE REVIEW

**Comment by bleskes** [comment]
@ywelsch this is now ready for review. If you have a moment it will be great.

File: docs/resiliency/index.asciidoc [hunk 1]
```diff
@@ -154,6 +140,20 @@ shard.
 
 == Completed
 
+[float]
+=== The _version field may not uniquely identify document content during a network partition (STATUS: DONE, v7.0.0)
```
Comment (ywelsch) [inline docs/resiliency/index.asciidoc]: Status: DONE. Can you prove it? ;)
Should we only move this after more extensive testing?

File: server/src/test/java/org/elasticsearch/versioning/SimpleVersioningIT.java [hunk 2]
```diff
@@ -213,75 +213,18 @@ public void testRequireUnitsOnUpdateSettings() throws Exception {
         }
     }
 
-    public void testInternalVersioningInitialDelete() throws Exception {
+    public void testCompareAndSEtInitialDelete() throws Exception {
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/versioning/SimpleVersioningIT.java]: s/SEt/Set/

**Review by ywelsch — APPROVED** [review]

File: docs/reference/migration/migrate_7_0/api.asciidoc [hunk 3]
```diff
@@ -2,6 +2,22 @@
 [[breaking_70_api_changes]]
 === API changes
 
+[float]
+==== Internal Versioning is no longer supported for optimistic concurrency control
+
+Elasticsearch maintains a numeric version field for each document it stores. That field
+is incremented by one with every change to the document. Until 7.0.0 the API allowed using
+that field for optimistic concurrency control, i.e., making a write operation conditional
+on the current document version. Sadly, that approach is flowed because the value of the
```
[the remark this answers is not in the document]
Comment (bleskes) [reply docs/reference/migration/migrate_7_0/api.asciidoc]: thx

File: docs/reference/migration/migrate_7_0/api.asciidoc [hunk 4]
```diff
@@ -2,6 +2,22 @@
...
+version doesn't always uniquely represent a change to the document. If a primary fails
+while handling a write operation, it may expose a version that will then be reused by the
+new primary.
+
+Due to that issue, internal versioning can no longer be used and is replaced by a new
+method based on sequence numbers. See <<optimistic-concurrency-control>> for more details.
+
+Note that the `EXTERNAL` versioning type is still fully supported.
```
[the remark this answers is not in the document]
Comment (bleskes) [reply docs/reference/migration/migrate_7_0/api.asciidoc]: fixed.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/WatcherRequestConvertersTests.java [hunk 5]
```diff
@@ -88,9 +88,11 @@ public void testPutWatch() throws Exception {
...
 
         if (randomBoolean()) {
-            long version = randomLongBetween(10, 100);
-            putWatchRequest.setVersion(version);
-            expectedParams.put("version", String.valueOf(version));
+            long seqNo = randomNonNegativeLong();
+            long ifPrimaryTerm = randomLongBetween(1, 200);
+            putWatchRequest.setIfSeqNo(seqNo);
+            expectedParams.put("if_seq_no", String.valueOf(seqNo));
+            expectedParams.put("if_primary_term", String.valueOf(ifPrimaryTerm));
```
[the remark this answers is not in the document]
Comment (bleskes) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/WatcherRequestConvertersTests.java]: good catch. I'll check why nothing failed.

[the remark this answers is not in the document]
Comment (bleskes) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/WatcherRequestConvertersTests.java]: confirmed - with enough runs tests fails.

**Comment by bleskes** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by bleskes** [comment]
@elasticmachine run elasticsearch-ci/default-distro

**Comment by jasontedor** [comment]
@elasticmachine test this please

**Comment by bleskes** [comment]
@elasticmachine run elasticsearch-ci/1

**Pull request merged**